### PR TITLE
DDF-3241 Added thread factory builder utility

### PR DESCRIPTION
**Pull request by mcalcote**
#### What does this PR do?
This adds the ability to add a thread factory to all blueprint executors, capture individual executor thread pools and configure them to use the general utility

#### Who is reviewing it? 
(please choose AT LEAST two reviewers that need to approve the PR before it can get merged)
@clockard @emmberk @rzwiefel @ahoffer @vinamartin @lessarderic

### CODE REVIEW

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java [hunk 1]
```diff
@@ -114,7 +115,8 @@
     private static ScheduledExecutorService createExecutorService() throws NumberFormatException {
         Integer threadPoolSize = NumberUtils.toInt(System.getProperty(
                 "org.codice.ddf.system.threadPoolSize"), THREAD_POOL_DEFAULT_SIZE);
-        return Executors.newScheduledThreadPool(threadPoolSize);
+        return Executors.newScheduledThreadPool(threadPoolSize,
+                StandardThreadFactoryBuilder.newThreadFactory("solrCloudClientFactory"));
```
Comment (emmberk) [inline platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java]: `solrCloudClientFactoryThread`?

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class StandardThreadFactoryBuilder implements ThreadFactory {
+
+    public static ThreadFactory newThreadFactory(String threadNamePrefix) {
+        return newThreadFactory(threadNamePrefix, true, Thread.NORM_PRIORITY);
+    }
+
+    public static ThreadFactory newThreadFactory(String threadNamePrefix, boolean isDaemon, int priority) {
```
Comment (rzwiefel) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: `@Nullable` threadNamePrefix?

File: platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java (hunk 1 above)
Comment (mcalcote) [reply platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrCloudClientFactory.java]: I just named it after the class.

File: catalog/core/catalog-core-impl/pubsub/src/main/java/ddf/catalog/pubsub/EventProcessorImpl.java [hunk 3]
```diff
@@ -74,7 +77,12 @@
 
     private Map<String, ServiceRegistration> existingSubscriptions;
 
-    private final ExecutorService threadPool = Executors.newCachedThreadPool();
+    private final ExecutorService threadPool = new ThreadPoolExecutor(0,
```
Comment (coyotesqrl) [inline catalog/core/catalog-core-impl/pubsub/src/main/java/ddf/catalog/pubsub/EventProcessorImpl.java]: ❓ Is there a reason you changed this to manual construction of your threadpool instead of just calling `Executors.newCachedThreadPool(StandardThreadFactoryBuilder.newThreadFactory("eventProcessorThread"))`

File: catalog/core/catalog-core-sourcemetricsplugin/src/main/java/ddf/catalog/metrics/source/SourceMetricsImpl.java [hunk 4]
```diff
@@ -307,7 +310,8 @@ public void addingSource(final Source source, Map props) {
         LOGGER.trace("ENTERING: addingSource");
 
         if (executorPool == null) {
-            executorPool = Executors.newCachedThreadPool();
+            executorPool = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60L, TimeUnit.SECONDS, new SynchronousQueue<>(),
```
Comment (coyotesqrl) [inline catalog/core/catalog-core-sourcemetricsplugin/src/main/java/ddf/catalog/metrics/source/SourceMetricsImpl.java]: ❓ Is there a reason you changed this to manual construction of your threadpool instead of just calling Executors.newCachedThreadPool(StandardThreadFactoryBuilder.newThreadFactory("eventProcessorThread"))

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java (hunk 2 above)
Comment (mcalcote) [reply platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: Prefix is misleading since I only tack a number on the end. Can update.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 5]
```diff
@@ -757,12 +760,22 @@ public void setQueryTypeNamespace(String queryTypeNamespace) {
 
     @Override
     public boolean isAvailable() {
+
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: ❓ Why this change at all? And even if necessary, it seems unrelated to the work at hand.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 6]
```diff
@@ -757,12 +760,22 @@ public void setQueryTypeNamespace(String queryTypeNamespace) {
...
+            sleep(30000);
+        } catch (InterruptedException e) {
+        }
         return availabilityTask.isAvailable();
     }
 
     @Override
     public boolean isAvailable(SourceMonitor sourceMonitor) {
         sourceMonitors.add(sourceMonitor);
+
```
Comment (coyotesqrl) [inline catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: ❓ Why this change at all? And even if necessary, it seems unrelated to the work at hand.

File: catalog/ui/search-ui/search-endpoint/src/main/java/org/codice/ddf/ui/searchui/query/controller/AbstractEventController.java [hunk 7]
```diff
@@ -57,7 +59,12 @@
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventController.class);
 
-    private final ExecutorService executorService = Executors.newCachedThreadPool();
+    private final ExecutorService executorService = new ThreadPoolExecutor(0,
```
Comment (coyotesqrl) [inline catalog/ui/search-ui/search-endpoint/src/main/java/org/codice/ddf/ui/searchui/query/controller/AbstractEventController.java]: ❓ Is there a reason you changed this to manual construction of your threadpool instead of just calling Executors.newCachedThreadPool(StandardThreadFactoryBuilder.newThreadFactory("eventProcessorThread"))

File: platform/metrics/platform-metrics-collector/src/main/java/ddf/metrics/collector/rrd4j/RrdJmxCollector.java [hunk 8]
```diff
@@ -237,7 +238,12 @@ public void init() {
         LOGGER.trace("ENTERING: init() for metric {}", metricName);
 
         if (executorPool == null) {
-            executorPool = Executors.newCachedThreadPool();
+            executorPool = new ThreadPoolExecutor(0,
```
Comment (coyotesqrl) [inline platform/metrics/platform-metrics-collector/src/main/java/ddf/metrics/collector/rrd4j/RrdJmxCollector.java]: ❓ Is there a reason you changed this to manual construction of your threadpool instead of just calling Executors.newCachedThreadPool(StandardThreadFactoryBuilder.newThreadFactory("eventProcessorThread"))

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        if (threadNamePrefix == null || threadNamePrefix.isEmpty()) {
+            threadNamePrefix = "thread";
+        }
+
+        return build(threadNamePrefix, isDaemon, priority);
+    }
+
+    private static ThreadFactory build(String threadNamePrefix, boolean isDaemon, int priority) {
+        final AtomicLong count = new AtomicLong(0);
```
Comment (coyotesqrl) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: ❗️ It's early on a Monday morning so forgive me if I'm missing something...but this is going to create a new `AtomicLong` on the stack every time you call `build`; you'll always return a thread with a suffix of `0`. Move it out of the method and make it a `static final` const.

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    public static ThreadFactory newThreadFactory(String threadNamePrefix, boolean isDaemon, int priority) {
+        if (threadNamePrefix == null || threadNamePrefix.isEmpty()) {
+            threadNamePrefix = "thread";
+        }
+
+        return build(threadNamePrefix, isDaemon, priority);
+    }
+
+    private static ThreadFactory build(String threadName, boolean isDaemon, int priority) {
+        final AtomicLong count = new AtomicLong(0);
```
Comment (coyotesqrl) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: _I'm re-adding this comment because it was lost when I posted my review._

❗️ It's early on a Monday morning so forgive me if I'm missing something...but this is going to create a new AtomicLong on the stack every time you call build; you'll always return a thread with a suffix of 0. Move it out of the method and make it a static final const.

Comment (mcalcote) [reply platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: I wrote this over the last few DevDays, so I wish I could say I knew what I was thinking. Possibly an artifact of a minor refactor?

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java (hunk 5 above)
Comment (mcalcote) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Most likely a typo or overzealous auto-format.

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java (hunk 6 above)
Comment (mcalcote) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Most likely a typo or overzealous auto-format.

File: catalog/core/catalog-core-impl/pubsub/src/main/java/ddf/catalog/pubsub/EventProcessorImpl.java (hunk 3 above)
Comment (mcalcote) [reply catalog/core/catalog-core-impl/pubsub/src/main/java/ddf/catalog/pubsub/EventProcessorImpl.java]: I *think* I was looking at the use of some of these threads and seeing if a different executor would provide a more optimized behavior. It seems that it ended up back where it started with the same call that a newCachedThreadPool would. Making the call you reference would provide the same effect in a more concise way. :+1:

File: catalog/core/catalog-core-sourcemetricsplugin/src/main/java/ddf/catalog/metrics/source/SourceMetricsImpl.java (hunk 4 above)
Comment (mcalcote) [reply catalog/core/catalog-core-sourcemetricsplugin/src/main/java/ddf/catalog/metrics/source/SourceMetricsImpl.java]: `newCachedThreadPool` will work here.

File: catalog/ui/search-ui/search-endpoint/src/main/java/org/codice/ddf/ui/searchui/query/controller/AbstractEventController.java (hunk 7 above)
Comment (mcalcote) [reply catalog/ui/search-ui/search-endpoint/src/main/java/org/codice/ddf/ui/searchui/query/controller/AbstractEventController.java]: `newCachedThreadPool` will work here.

File: platform/metrics/platform-metrics-collector/src/main/java/ddf/metrics/collector/rrd4j/RrdJmxCollector.java (hunk 8 above)
Comment (mcalcote) [reply platform/metrics/platform-metrics-collector/src/main/java/ddf/metrics/collector/rrd4j/RrdJmxCollector.java]: `newCachedThreadPool` will work here.

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java (hunk 10 above)
Comment (mcalcote) [reply platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: Looking at this again, I remember the idea was that all new thread pools/factories would have a count independent of each other. I suppose this doesn't provide much value in practice though, especially considering your point about the stack cost.

File: platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    public static ThreadFactory newThreadFactory(String threadName, boolean isDaemon,
+            int priority) {
+        if (threadName == null || threadName.isEmpty()) {
+            threadName = "thread";
+        }
+        return build(threadName, isDaemon, priority);
+    }
+
+    private static ThreadFactory build(String threadName, boolean isDaemon, int priority) {
+        return r -> {
```
Comment (coyotesqrl) [inline platform/util/platform-util/src/main/java/org/codice/ddf/platform/util/StandardThreadFactoryBuilder.java]: ✏️ I'm not sure if I misread this earlier and it was already correct or if the `AtomicLong` was incorrectly inside the factory method. 

Regardless, the right way to do this is to get rid of the global and create your new `AtomicLong` on line 35, right before your return statement. You can explicitly mark as final if you want, or just let Java treat it as effectively final for you.

**Comment by mcalcote** [comment]
build now

File: catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/IngestCommand.java [hunk 12]
```diff
@@ -208,7 +209,7 @@ protected Object executeWithSubject() throws Exception {
         final ArrayBlockingQueue<Metacard> metacardQueue = new ArrayBlockingQueue<>(
                 batchSize * multithreaded);
 
-        ExecutorService queueExecutor = Executors.newSingleThreadExecutor();
+        ExecutorService queueExecutor = Executors.newSingleThreadExecutor(StandardThreadFactoryBuilder.newThreadFactory("ingestCommandThread"));
```
[the remark this answers is not in the document]
Comment (mcalcote) [reply catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/catalog/IngestCommand.java]: Updated

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 13]
```diff
@@ -757,12 +760,20 @@ public void setQueryTypeNamespace(String queryTypeNamespace) {
...
+        }
         return availabilityTask.isAvailable();
     }
 
     @Override
     public boolean isAvailable(SourceMonitor sourceMonitor) {
         sourceMonitors.add(sourceMonitor);
+        try {
+            sleep(30000);
+        } catch (InterruptedException e) {
```
[the remark this answers is not in the document]
Comment (mcalcote) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Where did these changes come from?

File: catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java [hunk 14]
```diff
@@ -757,12 +760,20 @@ public void setQueryTypeNamespace(String queryTypeNamespace) {
 
     @Override
     public boolean isAvailable() {
+        try {
+            sleep(30000);
+        } catch (InterruptedException e) {
```
[the remark this answers is not in the document]
Comment (mcalcote) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: Removed

[the remark this answers is not in the document]
Comment (mcalcote) [reply catalog/spatial/csw/spatial-csw-source-common/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/common/source/AbstractCswSource.java]: I didn't see the change you originally commented on @coyotesqrl. This was definitely unintentionally caught up in these changes while debugging an unrelated issue.

**Comment by coyotesqrl** [comment]
❗️ Missed the thread pool in [`SourcePollerRunner`](https://github.com/codice/ddf/blob/master/catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePollerRunner.java#L102).
And while not especially worrisome, missed the pool in [`UnavailableUrls`](https://github.com/codice/ddf/blob/master/platform/util/platform-util-unavailableurls/src/main/java/org/codice/ddf/platform/util/http/UnavailableUrls.java#L52)

**Comment by coyotesqrl** [comment]
❗️ Additional locations that need thread factories:

1. [`ReliableResourceDownloader`](https://github.com/codice/ddf/blob/master/catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/resource/download/ReliableResourceDownloader.java#L265) - this one will be changed a bit shortly, as the initialization of that field in its declaration will be removed.
2. [blueprint for `registry-federation-admin-service-impl` module](https://github.com/codice/ddf/blob/master/catalog/spatial/registry/registry-federation-admin-service-impl/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L42)
3. `CrlChecker`, but this file is going to change considerably before your PR is merged.
4. [blueprint for `solr-appender`](https://github.com/codice/ddf/blob/master/platform/solr/solr-appender/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L19)

**Comment by coyotesqrl** [comment]
Also, for `UnavailableUrls`, you'll probably need to manually construct an inline `ThreadFactory` instead of using the `StandardThreadFactoryBuilder` because of the dependencies.

File: catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePollerRunner.java [hunk 15]
```diff
@@ -260,4 +260,4 @@ public String getOrganization() {
             return organization;
         }
     }
-}
+}
```
[the remark this answers is not in the document]
Comment (mcalcote) [reply catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePollerRunner.java]: Updated the executor being used here.

**Comment by mcalcote** [comment]
build now

**Comment by mcalcote** [comment]
This build failure was due to a timeout waiting for the remote bundle context in `TestFederation`. I didn't experience this while testing locally during development. Will investigate before running the CI build again.

**Comment by mcalcote** [comment]
build now

**Comment by mcalcote** [comment]
BFA has identified this failure with: `Connectivity issues to external sites`. I have manually confirmed that this itest class passes when not experiencing network issues.

**Comment by mcalcote** [comment]
build now

**Comment by mcalcote** [comment]
SolrCommands were missing a runtime dependency. Thanks @figliold for finding the issue!

**Comment by mcalcote** [comment]
build now

**Comment by mcalcote** [comment]
build now
